### ISSUE TITLE
bcachefs storage backend

### ISSUE THREAD

**Issue by FlyingWombat** [canonical/lxd#8960]
Please consider implementing a storage backend for [bcachefs](https://bcachefs.org/). Bcachefs is a new COW filesystem for Linux, that aims to improve upon the capabilities demonstrated by ZFS and Btrfs. More info on the motivation behind bcachefs can be found on the project's [Patreon funding page](https://www.patreon.com/bcachefs).

Bcachefs can currently be considered beta quality, and has not yet been mainlined into Linux. Still, it has a small community of active users, including myself. 

Here is the userspace code: https://github.com/koverstreet/bcachefs-tools

As of writing, snapshots and subvolumes are a WIP for bcachefs, but I don't think it is too soon to start on the storage backend for it.

**Comment by stgraber** [comment]
Yeah, I've been keeping an eye on bcachefs and it seems interesting, though realistically for us to spend the time needed to implement it, we're going to at least want it:

 - Merged in the mainline Linux kernel
 - At least one major distribution shipping it in a stable release
 - Packaged version of the tools and some kind of stable version of those
 - Support for the basics: snapshots, quotas, subvolumes (or at least clone of snapshots), ability to query disk usage per subvolume
 - Ideally a migration mechanism (send/receive)

It feels like we're still a few years away from having that at this point, but once that's the case, we'll be happy to assist anyone who wants to add it as a backend or may even put it on the team's roadmap and implement it ourselves.

Adding a new storage backend isn't all that difficult these days. We re-implemented some of ours in just 2-3 weeks worth of time. What is however very costly is the ongoing maintenance, bugfixes and testing that we do on all our storage backends. So before we add a new one, we're going to want to be sure that it will be stable and that it will have a reasonable userbase.

**Comment by FlyingWombat** [comment]
Update: bcachefs is now in mainline kernel v6.7 , released 2024-01-07.
It is supported by at least Archlinux and Nixos as of the v6.7 kernel release.
Bcachefs is feature equivalent with Btrfs and Zfs, as far as LXD is concerned -- with the exception of a dedicated subvolume send/receive mechanism.

**Comment by tomponline** [comment]
No plans for this at this time.
